harness/determined/common/detignore.py

```diff
@@ -0,0 +1,99 @@
+import os
+import pathlib
+from typing import TYPE_CHECKING, Callable, Iterable, List, Set
+
+if TYPE_CHECKING:
+    import pathspec
+
+from determined.common import constants, v1file_utils
+from determined.common.api import bindings
+
+
+def _build_detignore_pathspec(root_path: pathlib.Path) -> "pathspec.PathSpec":
+    ignore = list(constants.DEFAULT_DETIGNORE)
+    ignore_path = root_path / ".detignore"
+    if ignore_path.is_file():
+        with ignore_path.open("r") as detignore_file:
+            ignore.extend(detignore_file)
+
+    # Lazy import to speed up load time.
+    # See https://github.com/determined-ai/determined/pull/6590 for details.
+    import pathspec
+
+    return pathspec.PathSpec.from_lines(pathspec.patterns.GitWildMatchPattern, ignore)
+
+
+def make_shutil_ignore(root_path: pathlib.Path) -> Callable:
+    ignore_spec = _build_detignore_pathspec(root_path)
+
+    def _ignore(path: str, names: List[str]) -> Set[str]:
+        ignored_names = set()  # type: Set[str]
+        for name in names:
+            if name == ".detignore":
+                ignored_names.add(name)
+                continue
+
+            file_path = pathlib.Path(path) / name
+            file_rel_path = file_path.relative_to(root_path)
+
+            if (
+                file_path.is_dir() and ignore_spec.match_file(str(file_rel_path) + "/")
+            ) or ignore_spec.match_file(str(file_rel_path)):
+                ignored_names.add(name)
+
+        return ignored_names
+
+    return _ignore
+
+
+def os_walk_to_v1Files(
+    root_path: pathlib.Path, entry_prefix: pathlib.Path
+) -> Iterable[bindings.v1File]:
+    ignore_spec = _build_detignore_pathspec(root_path)
+
+    # We could use pathlib.Path.rglob for scanning the directory;
+    # however, the Python documentation claims a warning that rglob may be
+    # inefficient on large directory trees, so we use the older os.walk().
```

Comment:
This comment is ancient, but actually I think the fact that we skip directories is a much better reason to us os.walk.

I recommend just deleting the comment.